flash: use win32 wmi to try to find UF2 and COM ports when flashing

deadprogram opened the pull request:
This PR attempts to do for Windows 10 the same searching for default port and drive name when flashing boards as we are now doing for both Linux and macOS.

EDIT:
Both of these commands work:

```
tinygo flash -target circuitplay-express examples/blinky1
```

and

```
tinygo flash -target arduino-nano33 examples/blinky1
```

deadprogram commented:
Tested on Windows 10 using Surface Pro. Was able to `tinygo flash` both UF2 boards as well as an Arduino Nano33 using `bossac`.

One important note, is that you do have to double tap the reset button to put any of these boards into bootloader mode. That could be an additional refinement for a future PR. This one is done and ready for review.

aykevl commented on main.go:
```diff
@@ -209,7 +210,18 @@ func Flash(pkgName, port string, options *compileopts.Options) error {
+			var cmd *exec.Cmd
+			switch runtime.GOOS {
+			case "windows":
+				command := strings.Split(flashCmd, " ")
```
It's better to split commands using a dedicated command splitter, such as github.com/google/shlex, otherwise spaces may give invalid results.

deadprogram replied:
Was not aware of that package, just switched to using it.

deadprogram replied:
Hmm shlex does not appear to work as I expected, and now my `wmic` calls are failing. Have to look deeper into how it works, but my previous solution was working well enough.

Basically just separating the command from the params is probably enough, since what was happening was the runtime on Windows does some extra insertions of `\` characters.

Anyhow, I think I will keep the other changes, but go back to my simpler parsing that works for the Windows rules.

aykevl replied:
Just a little bit worried that future changes may trip something up, but I guess it's fine for now.

aykevl commented on main.go:
```diff
@@ -472,6 +502,29 @@ func flashHexUsingMSD(volume, tmppath string) error {
+func windowsFindUSBDrive() (string, error) {
+	cmd := exec.Command("wmic", "PATH", "Win32_LogicalDisk",
+		"get", "DeviceID,", "VolumeName,",
+		"FileSystem,", "DriveType")
```
A bit hacky but can't easily find a better way so OK.

aykevl commented on main.go:
```diff
@@ -505,6 +558,30 @@ func getDefaultPort() (port string, err error) {
 		portPath = "/dev/ttyACM*"
+	case "windows":
+		cmd := exec.Command("wmic", "PATH", "Win32_SerialPort",
+			"get", "DeviceID")
```
Did you see this?
http://a46554.blogspot.com/2018/05/go-lang-list-all-com-port-in-windows.html?m=1
Found it with a quick Google search. Not sure whether it works, but it might avoid an external command.

deadprogram replied:
We are already using the `wmic` command which is part of Windows itself. The command seems like the proper way to do this vs. search for some Windows registry entries that might change or even be entirely different based on system config.

aykevl commented on main.go:
```diff
@@ -438,9 +450,18 @@ func touchSerialPortAt1200bps(port string) error {
+		if err != nil {
+			return err
+		}
+		infoPath = path + "/INFO_UF2.TXT"
```
Looks like this only checks the first FAT device available (that may or may not be the right one).
That can possibly be fixed later, but could be surprising to new users if no device can be found while it's clearly plugged in (along a USB thumb drive).

deadprogram replied:
Added some additional logic to ensure that we only use attached volume with the correct name that matches the board, like we are doing on Linux/macOS.

aykevl replied:
I don't see the additional logic, but I do see that I missed the extra `volume` parameter to `windowsFindUSBDrive` which makes sure that only a volume with the correct name is returned. So that should be good.

deadprogram commented:
Pushed an additional commit modifying this PR to take into account some of @aykevl suggestions.

deadprogram commented:
Updated the previous commit to go back to the simpler parsing for Windows, but with enhancements for ensuring the correct devices are used.

deadprogram commented:
OK it took a bit to get the correct magic syntax to pass thru the code only called on Windows:
https://github.com/golang/go/blob/master/src/syscall/exec_windows.go#L17

Now it works exactly as expected, tested both gain and they worked perfectly.

Ready for merge!

aykevl commented on go.sum:
```diff
@@ -29,4 +29,4 @@ tinygo.org/x/go-llvm v0.0.0-20191103200204-37e93e3f04e2/go.mod h1:fv1F0BSNpxMfCL
 tinygo.org/x/go-llvm v0.0.0-20191113125529-bad6d01809e8/go.mod h1:fv1F0BSNpxMfCL0zF3M4OPFbgYHnhtB6ST0HvUtu/LE=
 tinygo.org/x/go-llvm v0.0.0-20191124211856-b2db3df3f257 h1:o8VDylrMN7gWemBMu8rEyuogKPhcLTdx5KrUAp9macc=
-tinygo.org/x/go-llvm v0.0.0-20191124211856-b2db3df3f257/go.mod h1:fv1F0BSNpxMfCL0zF3M4OPFbgYHnhtB6ST0HvUtu/LE=
+tinygo.org/x/go-llvm v0.0.0-20191124211856-b2db3df3f257/go.mod h1:fv1F0BSNpxMfCL0zF3M4OPFbgYHnhtB6ST0HvUtu/LE=
```
what's happening here? :thinking:

aykevl commented on main.go:
```diff
@@ -505,6 +558,30 @@ func getDefaultPort() (port string, err error) {
 		portPath = "/dev/ttyACM*"
+	case "windows":
+		cmd := exec.Command("wmic",
+			"PATH", "Win32_SerialPort", "WHERE", "Caption LIKE 'USB Serial%'", "GET", "DeviceID")
```
Ugh that looks a lot like SQL, but whatever...

deadprogram replied:
That is correct: https://docs.microsoft.com/en-us/windows/win32/wmisdk/wmi-and-sql

aykevl commented on main.go:
```diff
@@ -472,6 +502,29 @@ func flashHexUsingMSD(volume, tmppath string) error {
 
+func windowsFindUSBDrive(volume string) (string, error) {
+	cmd := exec.Command("wmic",
+		"PATH", "Win32_LogicalDisk", "WHERE", "VolumeName = '"+volume+"'",
```
There is no sanity-checking of the volume name. That's not a problem right now, but may be a problem if we want to accept untrusted volume names in the future.

deadprogram replied:
We can add that in a future PR.

deadprogram commented:
Went back and rebased to remove the change to `go.sum` which was just adding a new line at the end? Not sure where that came from, but I removed and rebased it away.

This is now ready to go, within the parameters of what it is trying to do.
